osu.Game/Rulesets/Mods/Mod.cs

```diff
@@ -197,6 +197,18 @@ public bool Equals(Mod other)
                            ModUtils.GetSettingUnderlyingValue(pair.Item2.GetValue(other))));
         }
 
+        public override int GetHashCode()
+        {
+            var hashCode = new HashCode();
+
+            hashCode.Add(GetType());
+
+            foreach (var (_, prop) in this.GetSettingsSourceProperties())
+                hashCode.Add(ModUtils.GetSettingUnderlyingValue(prop.GetValue(this)));
```

Comment:
I'm not sure we want to be doing this in `GetHashCode`. It is supposed to be a fast operation and this seems like it could have some overhead.

Reply:
Good point, I think we can have a static cache for these as they return the properties of the type, nothing to do with the instance itself.

Reply:
Why are you implementing in `GetHashCode` in the first place though?

Reply:
Because it is used in `DifficultyCacheLookup.GetHashCode()`, for equality comparisons during value retrieval I believe.

Reply:
Keep in mind `GetHashCode` is a best-effort return, not the ground truth. May want to read up on why it exists. May not be required in this case.

Reply:
`MemoryCachingComponent`s check for existence of computed values by doing `cache.TryGetValue(lookup, out var value)`, and the implementation of that method seems to require hash code to function properly:
![CleanShot 2021-08-17 at 18 19 16](https://user-images.githubusercontent.com/22781491/129753902-2b181c80-a37a-4638-a31a-05223b120ae9.png)

(doesn't consider equality even when `Equals` would return true)

I'll go with caching the properties just so that `GetHashCode()` doesn't get hindered by the overhead.

Reply:
Caching the properties doesn't eliminate all overhead. There's still the cost of fetching the values via reflection.

Looking at this from a different angle, I'm not really sure whether difficulty values for mods with settings *should* be cached at all. The reason why I'm doubting that is that I'm a bit skeptical as to what the hit rate of such a cache would be, especially for things like difficulty adjust where you have some hundreds of admissible *values* for each property, let alone *combinations* of values.

In other words, what I'm trying to put under consideration here is *not* caching difficulty values if they come from a mod with non-standard setting values. That would sidestep the issue entirely.

Reply:
It's not the value that's cached though? It's the property info itself, but yeah there's the overhead of fetching as well unfortunately.

Tempted to just leave this off but I'll need to benchmark because reflection is full of mysteries.

Reply:
Yeah seems there's still an overhead from the fetching alone:

- hash `GetType()` only (no properties fetching):
    |      Method | Times |        Mean |      Error |     StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |------------:|-----------:|-----------:|------:|------:|------:|----------:|
    | ModHashCode |     1 |    19.62 ns |   0.128 ns |   0.120 ns |     - |     - |     - |         - |
    | ModHashCode |    10 |   181.62 ns |   2.087 ns |   1.629 ns |     - |     - |     - |         - |
    | ModHashCode |   100 | 2,982.04 ns | 126.045 ns | 369.668 ns |     - |     - |     - |         - |

- hash `GetType()` and fetched properties without caching:
    |      Method | Times |        Mean |     Error |    StdDev |      Median |    Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |------------:|----------:|----------:|------------:|---------:|------:|------:|----------:|
    | ModHashCode |     1 |    55.11 us |  3.031 us |  8.842 us |    52.10 us |   2.5635 |     - |     - |      8 KB |
    | ModHashCode |    10 |   349.67 us | 29.659 us | 87.451 us |   295.46 us |  26.3672 |     - |     - |     82 KB |
    | ModHashCode |   100 | 2,729.53 us | 11.389 us |  8.891 us | 2,730.06 us | 265.6250 |     - |     - |    818 KB |

- hash `GetType()` and fetched properties with caching:
    |      Method | Times |       Mean |      Error |     StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |-----------:|-----------:|-----------:|--------:|------:|------:|----------:|
    | ModHashCode |     1 |   6.074 us |  0.2061 us |  0.5946 us |  0.1526 |     - |     - |     520 B |
    | ModHashCode |    10 |  42.969 us |  0.8216 us |  2.3173 us |  1.6479 |     - |     - |   5,200 B |
    | ModHashCode |   100 | 500.425 us | 16.0374 us | 45.4956 us | 16.1133 |     - |     - |  52,000 B |

I'll look at improving this, by once-off retrieving all bindable settings rather than fetching them every equality comparsion call.

Reply:
After caching all bindables from first retrieval, the results appear to be the following (`GetHashCode` with no properties added and `GetHashCode` with properties added, side to side):

|                  Method | Times |         Mean |      Error |       StdDev |       Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------ |------ |-------------:|-----------:|-------------:|-------------:|-------:|------:|------:|----------:|
| ModHashCodeNoProperties |     1 |     29.05 ns |   0.757 ns |     2.232 ns |     29.61 ns |      - |     - |     - |         - |
|             ModHashCode |     1 |    113.39 ns |   2.233 ns |     1.979 ns |    113.21 ns | 0.0203 |     - |     - |      64 B |
| ModHashCodeNoProperties |    10 |    359.92 ns |  25.999 ns |    75.014 ns |    339.14 ns |      - |     - |     - |         - |
|             ModHashCode |    10 |  2,262.31 ns | 159.937 ns |   464.006 ns |  2,101.73 ns | 0.2022 |     - |     - |     640 B |
| ModHashCodeNoProperties |   100 |  2,932.02 ns |  70.621 ns |   201.486 ns |  2,864.35 ns |      - |     - |     - |         - |
|             ModHashCode |   100 | 17,209.31 ns | 946.256 ns | 2,730.164 ns | 16,491.80 ns | 2.0294 |     - |     - |   6,400 B |

I'm still not sure what's causing the allocation when enumerating the settings and adding them to the hashcode.